jsonpb/jsonpb.go

```diff
@@ -644,7 +649,105 @@ func (u *Unmarshaler) UnmarshalNext(dec *json.Decoder, pb proto.Message) error {
 // permutations of the related Marshaler.
 func (u *Unmarshaler) Unmarshal(r io.Reader, pb proto.Message) error {
 	dec := json.NewDecoder(r)
-	return u.UnmarshalNext(dec, pb)
+	if err := u.UnmarshalNext(dec, pb); err != nil {
+		return err
+	}
+
+	return checkRequiredFields(pb)
+}
+
+// checkRequiredFields returns an error if any required field in the given proto message is not set.
+// This function is called from Unmarshal and assumes certain actions are done ahead.  While
+// required fields only exist in a proto2 message, a proto3 message can contain proto2 message(s).
+func checkRequiredFields(pb proto.Message) error {
+	// Most well-known type messages do not contain required fields.  The "Any" type may contain
+	// a message that has required fields.  When an Any message is being unmarshaled, the code will
+	// have invoked proto.Marshal on the embedded message to store the serialized message in
+	// Any.Value field, and that should have returned an error if a required field is not set.
+	// Hence skipping well-known types here.
+	if _, ok := pb.(wkt); ok {
+		return nil
+	}
+
+	v := reflect.ValueOf(pb).Elem()
```

Comment:
You will need to verify that pb is a pointer to a struct. If it isn't, then return.

Reply:
Done.  I'm guessing we'll simply ignore message if not pointer to struct.